File: python/day03/aoc3.py
import numpy as np  

with open("/Users/christopherauld/Desktop/AOC2018/data/day3/input.txt", "r") as file:
    s = [line.strip("\n") for line in file]

s2 = [t.split(" ") for t in s]

claims = {}
for c in s2:
    claims[int(c[0].replace("#", ""))] = {
        "left": int(c[2].split(',')[0]),
        "top": int(c[2].split(',')[1].strip(":")),
        "width": int(c[3].split('x')[0]),
        "height": int(c[3].split('x')[1]),
        "has_overlap": False
    }

fabric = np.full((1000, 1000), 0, dtype = int)
fabric_claimer = np.full((1000, 1000), 0, dtype = int)

def fill_grid(c, name):
    #start = np.array([c["left"], c["top"]])
    #end = np.array([c["width"], c["height"]]) + start
    for i in range(c["width"]):
        for j in range(c["height"]):
            fabric[j + c["top"], i + c["left"]] += 1
            if fabric_claimer[j + c["top"], i + c["left"]] == 0:
                fabric_claimer[j + c["top"], i + c["left"]] = name
            elif fabric_claimer[j + c["top"], i + c["left"]] != -1:
                claims[fabric_claimer[j + c["top"], i + c["left"]]]["has_overlap"] = True
                fabric_claimer[j + c["top"], i + c["left"]] = -1 
                c["has_overlap"] = True
            else:
                c["has_overlap"] = True


for key, item in claims.items():
    fill_grid(item, key)

for key, item in claims.items():
    if not item["has_overlap"]:
        # Part 2 answer
        print(key)

import matplotlib.pyplot as plt 
fabric[fabric <= 1] = 0
fabric[fabric > 1] = 1
plt.imshow(fabric)

count = 0
for row in fabric:
    for col in row:
        if col == 1:
            count +=1
# Part 1 answer
print(count)

plt.show()

File: 2018/go/day03/aoc3.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Claim struct {
	left   int
	top    int
	width  int
	height int
}

func main() {
	filename := "/Users/christopherauld/Desktop/AOC2018/data/day3/input.txt"
	b, err := os.Open(filename)
	if err != nil {
		fmt.Println(err)
	}

	// Part 1
	scanner := bufio.NewScanner(b)
	claims := make(map[int]Claim)
	claim_num := 0

	for scanner.Scan() {
		line := strings.Split(scanner.Text(), " ")
		claim_num, err = strconv.Atoi(strings.Replace(line[0], "#", "", -1))
		if err != nil {
			fmt.Println(err)
		}
		left, _ := strconv.Atoi(strings.Split(line[2], ",")[0])
		top, _ := strconv.Atoi(strings.Replace(strings.Split(line[2], ",")[1], ":", "", -1))
		width, _ := strconv.Atoi(strings.Split(line[3], "x")[0])
		height, _ := strconv.Atoi(strings.Split(line[3], "x")[1])
		claims[claim_num] = Claim{left, top, width, height}
	}

	var fabric [1000][1000]int
	//overlaps := make(map[int]bool)
	overlaps := []bool{}
	double_claimed := 0

	for i := 0; i <= claim_num; i++ {
		claim := claims[i]
		for w := claim.left; w < (claim.width + claim.left); w++ {
			for h := claim.top; h < (claim.height + claim.top); h++ {
				if fabric[h][w] == 0 {
					if i == 0 {
						fabric[h][w] = -2
					} else {
						fabric[h][w] = i
					}
				} else if fabric[h][w] > 0 || fabric[h][w] == -2 {
					if fabric[h][w] == -2 {
						overlaps[0] = true
					}
					overlaps[fabric[h][w]] = true
					overlaps[i] = true
					fabric[h][w] = -1
					double_claimed++
				} else {
					overlaps[i] = true
				}
			}
		}
	}
	fmt.Println("Part 1 = ", double_claimed)

	for i := 0; i <= claim_num; i++ {
		if !overlaps[i] {
			fmt.Println("Claim ", i, "has no overlap")
		}
	}
}
